### ISSUE TITLE
Derive the loss subclass from a metric class and add in an accuracy metric

### ISSUE THREAD

**Issue by finlaymorrison** [finlaymorrison/BunjiLearn#6]
There should be a parent abstract class 'Metric' which defines a pure virtual function for obtaining a value from a metric. The Loss class should derive from the 'Metric' class which then declares another pure virtual function for obtaining the derivative of the loss function. 

Effectively, differentiable metrics will be implemented as a loss function which can still be used as a model metric but can also be a training loss function.
